// @observablehq/framework/dist/javascript/imports.js
import { simple } from "acorn-walk";
import { isPathImport, relativePath, resolveLocalPath } from "../path.js";
import { getStringLiteralValue, isStringLiteral } from "./source.js";
import { syntaxError } from "./syntaxError.js";
function findExports(body) {
  const exports = [];
  simple(body, {
    ExportAllDeclaration: findExport,
    ExportNamedDeclaration: findExport
  });
  function findExport(node) {
    exports.push(node);
  }
  return exports;
}
function hasImportDeclaration(body) {
  let has = false;
  simple(body, {
    ImportDeclaration() {
      has = true;
    }
  });
  return has;
}
function findImports(body, path, input) {
  const imports = [];
  simple(body, {
    ImportDeclaration: findImport,
    ImportExpression: findImport,
    ExportAllDeclaration: findImport,
    ExportNamedDeclaration: findImport,
    CallExpression: findImportMetaResolve
  });
  function findImport(node) {
    const source = node.source;
    if (!source || !isStringLiteral(source))
      return;
    const name = decodeURI(getStringLiteralValue(source));
    const method = node.type === "ImportExpression" ? "dynamic" : "static";
    if (isPathImport(name)) {
      const localPath = resolveLocalPath(path, name);
      if (!localPath)
        throw syntaxError(`non-local import: ${name}`, node, input);
      imports.push({ name: relativePath(path, localPath), type: "local", method });
    } else {
      imports.push({ name, type: "global", method });
    }
  }
  function findImportMetaResolve(node) {
    const source = node.arguments[0];
    if (!isImportMetaResolve(node) || !isStringLiteral(source))
      return;
    const name = decodeURI(getStringLiteralValue(source));
    if (isPathImport(name)) {
      const localPath = resolveLocalPath(path, name);
      if (!localPath)
        throw syntaxError(`non-local import: ${name}`, node, input);
      imports.push({ name: relativePath(path, localPath), type: "local", method: "dynamic" });
    } else {
      imports.push({ name, type: "global", method: "dynamic" });
    }
  }
  return imports;
}
function isImportMetaResolve(node) {
  return node.callee.type === "MemberExpression" && node.callee.object.type === "MetaProperty" && node.callee.object.meta.name === "import" && node.callee.object.property.name === "meta" && node.callee.property.type === "Identifier" && node.callee.property.name === "resolve" && node.arguments.length > 0;
}
export {
  findExports,
  findImports,
  hasImportDeclaration,
  isImportMetaResolve
};
